fastcrypto/src/vrf.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::error::FastCryptoError;
+use crate::traits::AllowedRng;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::fmt::Debug;
+
+/// Represents a public key of which is use to verify outputs for a verifiable random function (VRF).
+pub trait VRFPublicKey: Serialize + DeserializeOwned + Eq + Debug {
+    type PrivateKey: VRFPrivateKey<PublicKey = Self>;
+}
+
+/// Represents a private key used to compute outputs for a verifiable random function (VRF).
+pub trait VRFPrivateKey: Serialize + DeserializeOwned + Eq + Debug {
+    type PublicKey: VRFPublicKey<PrivateKey = Self>;
+}
+
+/// A keypair for a verifiable random function (VRF).
+pub trait VRFKeyPair<const OUTPUT_SIZE: usize>: Serialize + DeserializeOwned + Eq + Debug {
+    type Proof: VRFProof<OUTPUT_SIZE, PublicKey = Self::PublicKey>;
+    type PrivateKey: VRFPrivateKey<PublicKey = Self::PublicKey>;
+    type PublicKey: VRFPublicKey<PrivateKey = Self::PrivateKey>;
+
+    /// Generate a new keypair using the given RNG.
+    fn generate<R: AllowedRng>(rng: &mut R) -> Self;
+
+    /// Generate a proof for the given input.
+    fn prove(&self, input: &[u8]) -> Self::Proof;
+
+    /// Compute both hash and proof for the given input.
+    fn output(&self, input: &[u8]) -> ([u8; OUTPUT_SIZE], Self::Proof) {
+        let proof = self.prove(input);
+        let output = proof.to_hash();
+        (output, proof)
+    }
+}
+
+/// A proof that the output of a VRF was computed correctly.
+pub trait VRFProof<const OUTPUT_SIZE: usize>: Serialize + DeserializeOwned + Eq + Debug {
+    type PublicKey: VRFPublicKey;
+
+    /// Verify the correctness of this proof.
+    fn verify(&self, input: &[u8], public_key: &Self::PublicKey) -> Result<(), FastCryptoError>;
+
+    /// Verify the correctness of this proof and VRF output.
+    fn verify_output(
+        &self,
+        input: &[u8],
+        public_key: &Self::PublicKey,
+        output: [u8; OUTPUT_SIZE],
+    ) -> Result<(), FastCryptoError> {
+        self.verify(input, public_key)?;
+        if self.to_hash() != output {
+            return Err(FastCryptoError::GeneralError);
+        }
+        Ok(())
+    }
+
+    /// Compute the output of the VRF with this proof.
+    fn to_hash(&self) -> [u8; OUTPUT_SIZE];
```

Comment:
I don't think the RFC allows it to not return a hash, and it's also the case for the other construction, RSA-FDH-VRF, that it will always return. In what ways do you think it can fail?

Reply:
I guess it's just a nit. I meant for general VRF functions, not only ones defined by the RFC.

Reply:
I see. I don't think it'll be too much trouble to change the output to a Result if we at some point implement one of those constructions.

Reply:
Are we going to resolve this and return a Result?

Reply:
If I understand Ben's like correctly, we're going to keep it as it is for now and only introduce the Result when needed.